### PR TITLE
Fetch kni-installer binary from a release payload

### PR DESCRIPTION
**Pull request by markmc**
Let's consume the installer the same way stock installer builds are intended to be consumed - by extracting it from the release payload. That way we're closer to being ready to transition to using stock openshift-install in future.

In order to do this, we need to produce a container image containing the installer, reference this image from a custom release payload, and drive the install (including installer extraction) from this custom payload.

---

- [x] Merge openshift/origin#22619 or make kni/release imagestream public
- [x] Add a custom version to the custom release payload using 'oc adm release new --name'
- [ ] Make it clear in the instructions what is needed when setting up a fresh namespace versus when using the existing 'kni' namespace
- [ ] Handle re-running the scripts where there is a previous installer build or release payload with the same version
- [ ] I've seen build_installer.sh time out waiting for the build - it has taken >15m

### CODE REVIEW

**Comment by markmc** [comment]
This PR needs openshift/origin#22619 and openshift-metalkube/kni-installer#53

**Comment by markmc** [comment]
We now also require openshift/origin#22636

**Comment by russellb** [comment]
Looks like the fix we depend on is merged, via https://github.com/openshift/origin/pull/22568

**Comment by russellb** [comment]
and the kni-installer patch is in.

One more patch needed still open: https://github.com/openshift/origin/pull/22619

**Comment by markmc** [comment]
> and the kni-installer patch is in.
> 
> One more patch needed still open: [openshift/origin#22619](https://github.com/openshift/origin/pull/22619)

This one is only needed because the `kni/release` is private - `ocp/release` must not be ... need to check that out

**Comment by eparis** [comment]
@smarterclayton look at what these guys are pulling off! You and I have gone round and round on this.

**Comment by cgwalters** [comment]
> Also, ultimately we need to get away from this way of caching a
pre-defined RHCOS image version - we instead need a way of caching
whatever version is driven by either openshift-install or machine
config upgrades.

i.e. https://github.com/openshift/installer/issues/1399

**Comment by markmc** [comment]
> > Also, ultimately we need to get away from this way of caching a
> > pre-defined RHCOS image version - we instead need a way of caching
> > whatever version is driven by either openshift-install or machine
> > config upgrades.
> 
> i.e. [openshift/installer#1399](https://github.com/openshift/installer/issues/1399)

Thanks, probably best discussed in openshift-metal3/kni-installer#37 - for bare-metal IPI, I don't think the user should have to do anything to download the image before kicking off the install

**Comment by stbenjam** [comment]
@markmc how do the build tags get set in this process? We need to build with ironic and libvirt.

**Comment by markmc** [comment]
> @markmc how do the build tags get set in this process? We need to build with ironic and libvirt.

Yep, see https://github.com/openshift-metal3/kni-installer/blob/master/images/baremetal/Dockerfile.ci

**Comment by markmc** [comment]
> > @markmc how do the build tags get set in this process? We need to build with ironic and libvirt.
> 
> Yep, see https://github.com/openshift-metal3/kni-installer/blob/master/images/baremetal/Dockerfile.ci

And `build_installer.sh` has:

```
      dockerfilePath: images/baremetal/Dockerfile.ci
```

**Comment by markmc** [comment]
I think this is ready to merge

There are a few improvements still listed in the checklist above, but ... well, we're relying on a custom payload now, so let's get this merged and make those improvements later if necessary

File: docs/release-payload.md [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+
+## Preparation and Configuration
+
+We build and publish within a namespace on an OpenShift
+cluster. First, prepare a `kubeconfig` with credentials to this
+cluster, and with the desired namespace set as the default:
+
+```
+$ oc --config=release-kubeconfig login https://api.ci.openshift.org --token=...
+$ oc --config=release-kubeconfig new-project kni
```
Comment (stbenjam) [inline docs/release-payload.md]: Do you need to run `new-project` every time?

File: docs/release-payload.md [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+$ oc --config=release-kubeconfig project kni
+$ oc --config=release-kubeconfig adm policy add-role-to-user admin <other admin>
```
Comment (stbenjam) [inline docs/release-payload.md]: Who is `<other admin>`? Is this to give someone else access?

File: docs/release-payload.md (hunk 1 above)
Comment (markmc) [reply docs/release-payload.md]: No, that's what "Make it clear in the instructions what is needed when setting up a fresh namespace versus when using the existing 'kni' namespace" in checklist refers to

File: docs/release-payload.md (hunk 2 above)
Comment (markmc) [reply docs/release-payload.md]: Yeah - e.g. I gave @russellb admin access (or he gave me, I forget)

**Comment by markmc** [comment]
> I haven't tried the process yet but I think it looks good. When rebasing the installer, is it expected to build the image and then test, or should we try to maintain some functionality in dev-scripts to build from git without a release payloud?

I'd be open to a dev hack to make testing the installer changes easier ... but shouldn't be the default path

**Comment by markmc** [comment]
Yeah, should just work with this PR

Let's hold off for now - the deploy I kicked off earlier failed. Will look into it

**Comment by derekhiggins** [comment]
Build SUCCESS, see build http://10.8.144.11:8080/job/dev-tools/644/

**Comment by markmc** [comment]
Sorry, didn't get back to this - if CI has passed, and one person sees it working for them locally ... then I'm fine with merging

**Comment by russellb** [comment]
I see that CI passed, but it's not working for me locally.  master does work for me.  Here's what I get:

```
+ oc adm release extract --registry-config installer--xMN0hzCslO/pullsecret --command=openshift-install --to installer--xMN0hzCslO registry.svc.ci.openshift.org/kni/release:4.1.0-rc.3-kni.1
error: unable to read image registry.svc.ci.openshift.org/kni/release:4.1.0-rc.3-kni.1: unauthorized: authentication required
```

**Comment by russellb** [comment]
> I see that CI passed, but it's not working for me locally. master does work for me. Here's what I get:
> 
> ```
> + oc adm release extract --registry-config installer--xMN0hzCslO/pullsecret --command=openshift-install --to installer--xMN0hzCslO registry.svc.ci.openshift.org/kni/release:4.1.0-rc.3-kni.1
> error: unable to read image registry.svc.ci.openshift.org/kni/release:4.1.0-rc.3-kni.1: unauthorized: authentication required
> ```

I got it to work by upgrading my client tools.  Script 01 ensures a 4.1 version of `oc` is installed, but it seems we need a new enough one?  I'm not sure how to detect that.  It could just a FAQ for a bit that people need to update `oc` ...

**Comment by stbenjam** [comment]
I saw same locally as well. Could we update `oc` client utils on every build?

Or just base it off the build date

```bash
OC_DATE=$(date -d $(oc version -o json  | jq -r '.clientVersion.buildDate') +%s)
[ $OC_DATE -ge 1558047075 ] || update_oc
```

**Comment by russellb** [comment]
> I saw same locally as well. Could we update `oc` client utils on every build?
> 
> Or just base it off the build date
> 
> ```shell
> OC_DATE=$(date -d $(oc version -o json  | jq -r '.clientVersion.buildDate') +%s)
> [ $OC_DATE -ge 1558047075 ] || update_oc
> ```

Yeah, I like that idea.  We know the current verison is new enough, at least.  Want to push that up as a follow-up PR we can merge right after this one?

**Comment by russellb** [comment]
OK, I've got a change based on Stephen's suggestion ready that I'm just going to push straight to the repo after merging this PR since we discussed it here.

**Pull request merged**